### PR TITLE
[AMD] Serve Kimi-K3 on gfx942: moonmath MLA multi-query verify, MXFP4 MoE, chunked prefix KV

### PR DESCRIPTION
**Pull request by tarik-sarac**
## Motivation

Kimi-K3 cannot be served on CDNA3 (MI300X / MI325X) today, and MLA models more
generally lose speculative decoding there. Four independent blockers, each of
which aborts or OOMs rather than degrading:

1. **`SituAndMul` does not build on ROCm.** `forward_cuda` JIT-builds
   `kimi_k3/situ_and_mul.cuh`, which includes `<cuda_fp8.h>` unguarded. Under
   hipcc the build fails *inside* CUDA-graph capture, so it surfaces only as
   `Capture cuda graph failed: ninja exited with status 1`.
2. **MXFP4 MoE aborts.** `aiter.fused_moe` has no gfx942 a4w4 path and dies in
   the activation quantiser with
   `fused_dynamic_mx_quant_moe_sort_hip: not support output type: fp4x2`.
3. **aiter's asm MLA has no kernel for 12 query heads** (Kimi-K3 at TP8 is
   96/8 = 12; aiter requires 4, 8, or a multiple of 16 in [16, 128]) and none
   past `qseqlen` 4, so a speculative draft window larger than 4 calls
   `abort()` during cuda-graph capture.
4. **Long-context prefill runs out of VRAM.** With a cached prefix the aiter MHA
   extend path re-gathers and up-projects the *entire* prefix through
   `kv_b_proj` in one shot. `AttnForwardMethod.MHA_CHUNKED_KV` already bounds
   this, but it was unreachable on ROCm: `maybe_disable_chunked_prefix_cache()`
   force-sets `disable_chunked_prefix_cache=True` for any backend absent from
   `CHUNKED_PREFIX_CACHE_SUPPORTED_ATTENTION_BACKENDS`.

This PR closes all four. It is stacked on #<PR1> (the single-query moonmath MLA
decode backend); the first commit here is that PR's commit rebased onto current
main.

## Modifications

**New attention backend — `--attention-backend moonmath_mla`**
(`layers/attention/moonmath_mla_backend.py`). Subclasses `AiterAttnBackend` and
takes over absorbed MLA with the [moonmath_attention](https://github.com/moonmath-ai/amd-kernels)
A16W8 kernels (bf16 Q / fp8 KV), which read SGLang's existing fused-576
`MLATokenToKVPool` key buffer directly — page_size 1, device-driven, cuda-graph
safe. Two shapes, both `H <= 16`:

| forward mode | kernel |
| --- | --- |
| decode, `q_len` 1 | `mla_decode_a16w8_paged_dev` |
| `TARGET_VERIFY`, `q_len` 4..8 | `mla_decode_a16w8_multiq_paged_dev` |

Everything else — prefill, bf16 KV, any unsupported geometry — falls through to
`AiterAttnBackend`, so the backend is a strict superset of aiter's behaviour.
The multi-query arm keeps the draft window resident per CTA and applies
end-aligned causal masking inside it, which is what makes speculative decoding
reachable at all past a window of 4. Q stays bf16 in both arms, so there is no
query scale to calibrate and the verify logits are not perturbed.

**Head padding for the fallback paths.** The paths this backend does not serve
itself still run aiter's kernel, so Q is zero-padded to aiter's 16-head minimum
for them. Unconditional at the head counts aiter has no kernel for, because
there the current behaviour is a process abort — there is no prior result to
preserve. The pad is on the query side only and MLA has no reduction across the
head axis, so the first `H` output rows are bit-identical to an unpadded call.
`AiterAttnBackend` gains one hook, `skip_mla_head_count_assert`, so a subclass
serving MLA with its own kernel is not rejected at construction.

**Kimi-K3 on gfx942.** A `SituAndMul.forward_hip` override onto the Triton twin
SGLang already ships (arithmetically identical, still one fused kernel), and a
gfx942 MXFP4 MoE route through aiter's *Triton* MXFP4 GEMMs
(`moe_runner/aiter_mxfp4_triton.py`), which lower `tl.dot_scaled` onto the bf16
MFMA CDNA3 has always had. `moe_runner/mxfp4_situ_fused.py` adds a SiTU variant
of aiter's fused SwiGLU epilogue so the routed experts keep their activation
fused into the GEMM. Both are behind `use_triton_mxfp4_moe()`, which is False on
every other arch and on any build where the Triton kernels do not import, so no
other MoE path changes.

**Split-KV verify on gfx942** (`kernels/ops/attention/verify_splitkv.py`,
`triton_backend.py`). Enabled on gfx942 as well as gfx95, and the split cap is
now CU-aware: `extend_attention_fwd` walks the prefix serially under a
`(bs, h_q, 1)` grid, so a per-TP-rank draft shard leaves most of the device
idle, and the flat `MAX_N_SPLITS = 16` was tuned at a base grid that already
fills the device. When the base grid underfills it, the cap is raised by the
shortfall — bounded at 128 splits, and bounded again by a scratch budget so a
large `max_bs` cannot balloon the stage-1 partials. **Shapes that already fill
the device keep their tuned value bit-identically** (`choose_n_splits` with no
`base_programs` is unchanged).

**Chunked prefix KV on the aiter / moonmath_mla backends.** Adds both to
`CHUNKED_PREFIX_CACHE_SUPPORTED_ATTENTION_BACKENDS`, returns
`MHA_CHUNKED_KV` from `handle_attention_aiter` once the cached prefix passes
`chunked_prefix_cache_threshold`, and implements
`AiterAttnBackend._forward_extend_mha_chunk()`. Two ROCm-specific fixes were
needed to make that path executable: `sgl_kernel` exports a `merge_state_v2`
wrapper on ROCm but never registers the op, so the CUDA-gated import left the
name undefined and the chunked path died with `NameError` (now uses
`merge_state`, which already dispatches to the Triton kernel on AMD); and
aiter's `flash_attn_varlen_func` returns the softmax LSE heads-first
`[num_heads, num_tokens]` while `merge_state` indexes `[num_tokens, num_heads]`.

**Environment surface: one new variable.**
`SGLANG_MOONMATH_MLA_MULTIQ_VERIFY` (default **on**) routes `TARGET_VERIFY`
through the multi-query kernel; set it to 0 to force the aiter path for A/B.

Serving Kimi-K3 with an fp8 KV cache and DSpark at block size 7 (verify window
8, exactly the multi-query kernel's range):

```bash
SGLANG_USE_AITER=1 python3 -m sglang.launch_server \
  --tp 8 --model moonshotai/Kimi-K3 \
  --attention-backend moonmath_mla --kv-cache-dtype fp8_e4m3 \
  --speculative-algorithm DSPARK \
  --speculative-draft-model-path RadixArk/Kimi-K3-DSpark \
  --speculative-dspark-block-size 7 --trust-remote-code
```

## Accuracy Tests

**MLA decode and multi-query verify vs an fp32 reference** — per-TP-rank shapes
on one MI300X, driving the backend's real `forward_decode` / `forward_extend`
(gates, seq_lens staging, kv-split planning and operand splitting included)
against an fp32 reference computed on the dequantized fp8 KV pool:

| case | shapes | rel L2 |
| --- | --- | --- |
| Kimi-K3 decode (H=12) | bs 1/4/16/32, S 2k–8k | 2.7e-3 |
| Kimi-K3 multi-query verify (H=12) | q_len 4,5,6,7,8 × bs 1/4/8/16 | 2.6e-3 |
| Kimi-K3 ragged verify (H=12) | spans `[4096, 517, 8192, 1031]`, `[2048, 2049, 33, 9000, 128, 4096]` | 2.7e-3 |
| DeepSeek-V3 decode + verify (H=16) | bs 8, S 4k | 2.7e-3 |

2.7e-3 is the fp8 KV quantization floor for these shapes, not kernel error — the
reference dequantizes the same fp8 bytes. The gates were separately checked to
decline (fall back to aiter) at `q_len` 3, `q_len` 9, `H = 32`, and with the
multi-query knob off. 20/20.

**Chunked prefix KV** — `test/registered/kernels/test_aiter_mha_chunked_kv.py`
asserts chunk + merge reproduces a single-shot attention over the whole context
at Kimi-K3-at-TP8 shapes, including an uneven final chunk. Tolerance 5e-3 is the
bf16 reassociation floor from splitting the KV stream.

**Unit tests** — `test/registered/unit/.../test_moonmath_mla_backend.py`, 13
tests: eligibility gating (head count, KV dtype, forward mode, MLA dims,
non-MLA), the multi-query window bounds, and A16W8 decode vs an fp32 reference
on GPU. All green.

## Speed Tests and Profiling

Split-KV verify on gfx942, MI300X, head_dim 64, h_q 8, h_kv 2, q_len 4, 250k
prefix: bf16 7.63 → 0.26 ms, fp8 9.30 → 0.23 ms. Accuracy against an exact fp32
reference is unchanged; the two paths differ only in reduction order.

Chunked prefix KV, 8× MI325X, Kimi-K3 TP8, 937k-token cold prefill:

| | before | after |
| --- | --- | --- |
| free VRAM at peak | 0.02 GiB | 17.90 GiB |
| transient | 19.3 GiB | 1.99 GiB |
| `kv_b_proj` GEMM M | 394564 | 32768 |

Peak becomes context-independent (237.86 GB at 400k vs 237.79 GB at 937k) for
+1% prefill latency, which is the per-chunk LSE merges. Without it the scheduler
aborted with `HSA_STATUS_ERROR_OUT_OF_RESOURCES` three times in 45 minutes at
347k–411k context.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### Notes for reviewers

- The backend requires `moonmath-attention` (branch `mla-decode-a16w8-multiq`)
  and `--kv-cache-dtype fp8_e4m3`. It is opt-in via `--attention-backend`, and
  the import is at backend construction, so no other configuration is affected.
- All new gfx942 code is arch-gated or backend-gated; the diff removes only 9
  lines from existing files, each forced by a behaviour change
  (`choose_n_splits`' signature and cap, the aiter MLA head-count assert, the
  gfx95-only split-KV gate, and the `kimi_k3_vl` aiter conv2d import).
- Both new tests are registered for the AMD CI suite
  (`stage-b-test-1-gpu-small-amd`).

---
### CI States

Latest PR Test (Base): :x: [Run #31212470108](https://github.com/sgl-project/sglang/actions/runs/31212470108)
Latest PR Test (Extra): :x: [Run #31212471175](https://github.com/sgl-project/sglang/actions/runs/31212471175)

### CODE REVIEW

**Comment by HaiShaw** [comment]
cc @amd-danli103

**Comment by amd-danli103** [comment]
hi @tarik-sarac, thanks for the thorough writeup — the motivation for each of these is clear. Before we review in depth, may I ask you to split this up? Not necessarily by commit — how finely you slice it is up to you; if some of these are more natural together given how you've been developing them, that's fine by me.

1. K3 on gfx942: `SituAndMul.forward_hip`, Triton MXFP4 MoE, `kimi_k3_vl` import fix (c07fedc). Would need an accuracy test for the ~890 lines of new MXFP4 MoE kernels, which the current description doesn't cover.
2. `verify_splitkv` — the gfx942 gate and the CU-aware split cap, together with the `intra_block_causal` / `draft_extend_v2` change in c66b5bc, which the description doesn't mention.
3. Chunked prefix KV on the aiter backend (415afb5), including the ROCm `merge_state_v2` fix in `forward_mha.py`.
4. The `moonmath_mla` backend and multi-query verify.

The main reason is that 1–3 have no external dependency and can be validated in CI today, while 4 needs `moonmath-attention` from a git branch (it isn't on PyPI) and so can't be exercised upstream at all. Bundled together, the parts we
could merge now are blocked on the part we can't verify.

Happy to review each one as it lands.

@HaiShaw does splitting it along these lines sound right to you?

**Comment by tarik-sarac** [comment]
Thanks for your time and review @amd-danli103, split into four as suggested:

| PR | |
| --- | --- |
| #35525 | [AMD] MXFP4 MoE on gfx942 through aiter's Triton kernels |
| #35521 | [AMD] Split-KV verify on gfx942: CU-aware split cap, bidirectional draft blocks |
| #35522 | [AMD] Enable chunked prefix KV on the aiter backend |
| #35523 | [AMD] moonmath MLA attention backend |

The accuracy test you asked for is in #35525 —
`test/registered/moe/test_aiter_mxfp4_triton_moe.py`, both new kernels against
an fp32 reference that dequantizes the same MXFP4 bytes, running against a stock
aiter. The `intra_block_causal` / `draft_extend_v2` change is described in
#35521.
